man/tag.1
.TH TAG "1" "November 2010" "" "User Commands"

.SH NAME
tag \- tag and search for files with keywords

.SH SYNOPSIS
.B tag
\fI-t\fR [[\fIMODIFIER\fR]\fIKEYWORD\fR]... \fIFILE\fR|\fIDIRECTORY\fR
.br
.B tag
\fI-f\fR [[\fIMODIFIER\fR]\fIKEYWORD\fR]... \fIDIRECTORY\fR
.br
.B tag
\fI-q\fR \fIFILE\fR|\fIDIRECTORY\fR
.br
.B tag
\fILONG-OPTION\fR

.SH DESCRIPTION
.\" Add any additional description here
.TP
\fB\-t\fR
tag a file or directory with the specified keywords
.TP
\fB\-f\fR
search for files and directories that have been tagged with the specified keywords
.TP
\fB\-q\fR
query what keywords are associated with a file or directory
.TP
\fB\-\-help\fR
display this help and exit
.TP
\fB\-\-version\fR
output version information and exit

.SH KEYWORD MODIFIERS
When performing file tagging and searching, keywords can optionally be prefixed by a modifier character. These modifiers will change how 
.B tag 
interprets the keyword for the specified operation.

.SS "File Tagging"
.TP
.B +KEYWORD (or just KEYWORD)
Will add the keyword
.B KEYWORD 
to the specified file only if it has not already been added.
.TP
.B -KEYWORD
Will remove the keyword
.B KEYWORD 
from the specified file only if it has already been added.
.TP
.B :KEYWORD
Will remove any keyword that contains
.B KEYWORD 
as a substring.
.TP
.B .KEYWORD
Will remove any keyword that does not contains
.B KEYWORD 
as a substring.

.SS "File Searching"
.TP
.B +KEYWORD (or just KEYWORD)
Will return files that contain the keyword
.BR "KEYWORD"
.TP
.BR -KEYWORD
Will filter out returned files that contain the keyword
.BR "KEYWORD"
.TP
.BR :KEYWORD
Will return files that contain the keyword
.BR "KEYWORD" 
as a substring.
.TP
.BR .KEYWORD
Will filter out returned files that contain the keyword
.BR "KEYWORD" 
as a substring.

.SS "Useful Tricks"
.TP
.B "tag -f : DIRECTORY"
Will return all files that have been tagged, inside of 
.B DIRECTORY
.TP
.B "tag -f + DIRECTORY"
Will return all files that can hold more tags, inside of 
.B DIRECTORY
.TP
.B "tag -f : - DIRECTORY"
Will return all files that cannot hold more tags, inside of 
.B DIRECTORY
.TP
.B "tag -t : TARGET"
Will remove all keywords associated with 
.B TARGET

.SH AUTHOR
Written by Josh Komoroske.
.br
tag source code repository: <http://github.com/sigint/tag>

.SH COPYRIGHT
Copyright \(co 2010 Josh Komoroske
.br
License BSD:
.br
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL JOSHUA KOMOROSKE BE LIABLE FOR ANY
DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



src/tag.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <errno.h>
#include <sys/stat.h>
#include "filesys.h"
#include "tagdb.h"
#include "res.h"

void search(const char *,const char **,int);
int query_file(const char*);




int main(int argc,char **argv){

	if(argc<2){
		fprintf(stderr,"tag: mising operand\n");
		fprintf(stderr,"Try `tag --help' for more information.\n");
		return 0;
	}
	
	if(!strcmp(argv[1],"--help")){
		printf(helptext);
		return 0;
	}
	
	if(!strcmp(argv[1],"--version")){
		printf(versiontext,VERSION_MAJOR,VERSION_MINOR,VERSION_BUILD);
		return 0;
	}

	if(argc<3){
		fprintf(stderr,"tag: mising operand\n");
		fprintf(stderr,"Try `tag --help' for more information.\n");
		return 0;
	}

	if(!strcmp(argv[1],"-t")){//tag a file
			tag_tagfile(argv[argc-1],(const char **)argv+2,argc-3);

	}else if(!strcmp(argv[1],"-f")){//find file(s)
		//printf("find mode\n");
		struct stat s;
		if(lstat(argv[argc-1],&s)){
			fprintf(stderr,"tag: `%s\': No such directory\n",argv[argc-1]);
			return 0;
		}

		if(!S_ISDIR(s.st_mode)){
			fprintf(stderr,"tag: `%s\': Not a directory\n",argv[argc-1]);
			return 0;
		}

			int len=strlen(argv[argc-1]);
			char *path=malloc((len+4)*sizeof(*path));
			strcpy(path,argv[argc-1]);
			if(path[len-1]!='/'){
				strcat(path,"/");
			}
			search(path,(const char **)argv+2,argc-3);
			free(path);path=NULL;

	}else if(!strcmp(argv[1],"-d")){//dump file tags
		//dump_tagfile(argv[argc-1]);
		

	}else if(!strcmp(argv[1],"-q")){//query file tags
		query_file(argv[argc-1]);

	}else{
		fprintf(stderr,"tag: invalid operand\n");
		fprintf(stderr,"Try `tag --help' for more information.\n");
	}


	return 0;
}




int query_file(const char *target){

	char *path,*name,*tagfile;
	if(!extract_paths(target,&path,&name,&tagfile)){
		return 0;
	}
	
	FILE *ftags;
	if((ftags=fopen(tagfile,"rb"))==NULL){return 0;}

	query_tagfile(name,ftags);

	fclose(ftags);
	free(name);name=NULL;
	free(path);path=NULL;
	free(tagfile);tagfile=NULL;

	return 1;
}




#ifdef _DIRENT_HAVE_D_TYPE
	#ifndef __USE_BSD
		#define DT_UNKNOWN	0
		#define DT_FIFO	    1
		#define DT_CHR		  2
		#define DT_DIR		  4
		#define DT_BLK		  6
		#define DT_REG		  8
		#define DT_LNK		 10
		#define DT_SOCK	   12
		#define DT_WHT		 14
	#endif




void search(const char *path,const char **tags,int tagc){
	struct dirent *dp;
	DIR *dir;

	if((dir=opendir(path))==NULL){return;}

	while((dp=readdir(dir))!=NULL){

		if(dp->d_type==DT_REG){//if it's a regular file
			if(strcmp(dp->d_name,TAGFILE_FILENAME))continue;

			char *temp=malloc((strlen(path)+strlen(dp->d_name)+1)*sizeof(*temp));//lots of wasted cycles
			sprintf(temp,"%s%s",path,dp->d_name);
			FILE *ftags;
			if((ftags=fopen(temp,"rb"))!=NULL){search_tagfile(path,tags,tagc,ftags);}
			fclose(ftags);
			free(temp);

		}else if(dp->d_type==DT_DIR){//if it's a directory
			if(!strcmp(dp->d_name,".") || !strcmp(dp->d_name,".."))continue;

			char *temp=malloc((strlen(path)+strlen(dp->d_name)+2)*sizeof(*temp));//lots of wasted cycles
			sprintf(temp,"%s%s/",path,dp->d_name);
			search(temp,tags,tagc);
			free(temp);

		}

	}
	closedir(dir);
}




#else




void search(const char *path,const char **tags,int tagc){



	struct dirent *dp;
	//struct stat s;
	DIR *dir;

	if((dir=opendir(path))==NULL){return;}

	while((dp=readdir(dir))!=NULL){

		if(!strcmp(dp->d_name,".") || !strcmp(dp->d_name,".."))continue;
	
		char *temp=malloc((strlen(path)+strlen(dp->d_name)+4)*sizeof(*temp));//lots of wasted cycles (;_;)
		sprintf(temp,"%s%s",path,dp->d_name);

		lstat(temp,&s);

		if(S_ISREG(s.st_mode)){//if it's regular file
			if(!strcmp(dp->d_name,TAGFILE_FILENAME)){
				FILE *ftags;
				if((ftags=fopen(temp,"rb"))!=NULL){search_tagfile(path,tags,tagc,ftags);}
				fclose(ftags);

			}
		}else if(S_ISDIR(s.st_mode) && !S_ISLNK(s.st_mode)){//if it's a directory
				strcat(temp,"/");
				search(temp,tags,tagc);
		}

		free(temp);


	}
	closedir(dir);


}

#endif

